shards/watcher.go

```diff
@@ -55,9 +55,11 @@ func NewDirectoryWatcher(dir string, loader shardLoader) (io.Closer, error) {
 		loader:     loader,
 		quit:       quitter,
 	}
-	if err := sw.scan(); err != nil {
-		return nil, err
-	}
+	go func() {
```

Comment:
Serving search and list requests (i.e. starting the web-server) before loading all shards up will cause the loading to deadlock since it won't be able to acquire the write locks necessary to update the shards in `shardedSearcher`, because all the concurrent read requests on-going.

I just tested this hypothesis in our `zoekt-load-test` VM by running a load test against the web-server at 10 requests per second while it was loading up the shards. The result matches my suspicion: The logs for loading up a shard stop being written and after a timeout, there's a panic.

```
root@zoekt-load-test:~# env GOGC=50 GODEBUG=gctrace=1 ./zoekt-webserver-async -index /data/index -pprof
...
2019/07/12 09:02:31 loading /data/index/gitolite.sgdev.org%2F13k%2Fdylanaraps%2Fwal_v15.00000.zoekt
2019/07/12 09:02:31 loading /data/index/ghe.sgdev.org%2Fsourcegraph%2Fapachecn-xgboost-doc-zh_v15.00000.zoekt
2019/07/12 09:02:31 loading /data/index/gitolite.sgdev.org%2F13k%2FOpenSCAP%2Fcontainer-compliance_v15.00000.zoekt
2019/07/12 09:02:31 loading /data/index/gitolite.sgdev.org%2F13k%2Fignamiguel%2Falgo4cobol_v15.00000.zoekt
2019/07/12 09:02:31 loading /data/index/ghe.sgdev.org%2Fsourcegraph-python%2Felastic-elasticsearch-dsl-py_v15.00000.zoekt
2019/07/12 09:02:31 loading /data/index/ghe.sgdev.org%2Fsourcegraph%2Fsloria-cookiecutter-flask_v15.00000.zoekt
2019/07/12 09:02:31 loading /data/index/ghe.sgdev.org%2Fsourcegraph-r%2Fthomasp85-patchwork_v15.00000.zoekt
2019/07/12 09:02:31 loading /data/index/ghe.sgdev.org%2Fsourcegraph%2Fdbcli-pgcli_v15.00000.zoekt
gc 36 @3.245s 13%: 2.5+182+2.3 ms clock, 163+11/1909/0+152 ms cpu, 522->548->463 MB, 576 MB goal, 64 P
gc 37 @4.083s 11%: 4.4+145+1.6 ms clock, 286+60/1991/0.58+105 ms cpu, 646->673->557 MB, 694 MB goal, 64 P
gc 38 @5.149s 9%: 2.1+99+3.6 ms clock, 137+129/993/0.68+233 ms cpu, 793->825->657 MB, 836 MB goal, 64 P
gc 39 @6.117s 8%: 2.2+207+4.2 ms clock, 143+116/1077/0+274 ms cpu, 943->989->817 MB, 986 MB goal, 64 P
gc 40 @7.056s 7%: 0.50+312+1.9 ms clock, 32+232/1114/11+124 ms cpu, 1174->1229->1004 MB, 1225 MB goal, 64 P
gc 41 @8.599s 6%: 1.9+310+2.3 ms clock, 124+119/1303/0.073+152 ms cpu, 1445->1487->1220 MB, 1507 MB goal, 64 P
gc 42 @10.688s 5%: 1.3+142+10 ms clock, 86+322/1123/20+673 ms cpu, 1780->1843->1513 MB, 1830 MB goal, 64 P
gc 43 @13.226s 4%: 0.68+172+2.5 ms clock, 43+123/1233/0+164 ms cpu, 2190->2235->1846 MB, 2269 MB goal, 64 P
gc 44 @16.385s 4%: 2.0+1116+1.3 ms clock, 133+147/2401/0+84 ms cpu, 2697->2743->2260 MB, 2769 MB goal, 64 P
gc 45 @24.192s 3%: 2.3+483+7.3 ms clock, 152+635/1577/0+469 ms cpu, 3323->3390->2806 MB, 3390 MB goal, 64 P
gc 46 @32.143s 2%: 5.9+127+6.6 ms clock, 378+205/1528/0+425 ms cpu, 4133->4178->3469 MB, 4209 MB goal, 64 P
2019/07/12 09:03:01 reloading: /data/index/ghe.sgdev.org%2Fsourcegraph-javascript%2Fyeoman-generator-angular_v15.00000.zoekt, err invalid argument
gc 47 @37.864s 2%: 3.9+114+1.3 ms clock, 250+259/1304/0+83 ms cpu, 5119->5167->4252 MB, 5203 MB goal, 64 P
2019/07/12 09:03:06 reloading: /data/index/ghe.sgdev.org%2Fsourcegraph-cpp%2Fjolibrain-deepdetect_v15.00000.zoekt, err invalid argument
2019/07/12 09:03:11 reloading: /data/index/ghe.sgdev.org%2Fsourcegraph-javascript%2Fmdbootstrap-bootstrap-material-design_v15.00000.zoekt, err invalid argument
gc 48 @44.547s 1%: 0.73+278+1.1 ms clock, 47+192/1746/0.45+72 ms cpu, 6273->6329->5242 MB, 6379 MB goal, 64 P
gc 49 @53.687s 1%: 2.3+1268+5.9 ms clock, 147+2950/4018/0+383 ms cpu, 7736->7851->6554 MB, 7863 MB goal, 64 P
2019/07/12 09:03:27 watchdog: Get http://:6070: context deadline exceeded
panic: watchdog: Get http://:6070: context deadline exceeded

goroutine 15 [running]:
log.Panicf(0x8bd156, 0xc, 0xc0f0605f90, 0x1, 0x1)
        /home/tomas/sdk/go1.12.4/src/log/log.go:340 +0xc0
main.watchdog(0x6fc23ac00, 0xc000024770, 0xc)
        /home/tomas/Code/zoekt/cmd/zoekt-webserver/main.go:263 +0x1b3
created by main.main
        /home/tomas/Code/zoekt/cmd/zoekt-webserver/main.go:210 +0x92d
```

I don't think we should start the web-server before all shards are loaded. Even if we fixed the concurrency control issues above, would we want partial results to be served while we're still loading up shards?

We can try to make loading up shards faster. [This commit](https://github.com/sourcegraph/zoekt/pull/11/commits/eef52eda69f868ab0ec59476e20a90e040678148) in a PR I have open improves things by reducing contention, but hasn't done anything to speed up individual shard loading.

Reply:
> would we want partial results to be served while we're still loading up shards?

As it stands, indexed search being unavailable => we use searcher. Atleast if we have partial index we don't hit searcher for everything.

Reply:
> Atleast if we have partial index we don't hit searcher for everything.

What makes this undesirable, specially for a short period of time?

Reply:
> What makes this undesirable, specially for a short period of time?

I think it depends on how short, the short period of time is. I suppose either way setting a goal of loading all shards quickly is a good thing and if we can improve it we can remove the need for async. So yeah, lets go down that path first before considering async loading.

Reply:
@keegancsmith, in that case it sounds like this PR is no longer needed.